### PR TITLE
[IRGen] Re-introduce TypeLayout strings

### PR DESCRIPTION
**Pull request by drexin**
Layout strings encode the structure of a type into a byte string that can be
interpreted by a runtime function to achieve a destroy or copy. Rather than
generating ir for a destroy/assignWithCopy/etc, we instead generate a layout
string which encodes enough information for a called runtime function to
perform the operation for us. Value witness functions tend to be quite large,
so this allows us to replace them with a single call instead. This gives us the
option of making a codesize/runtime cost trade off.

### CODE REVIEW

**Comment by drexin** [comment]
https://github.com/apple/swift-syntax/pull/1073

@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
https://github.com/apple/swift-syntax/pull/1073

@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
https://github.com/apple/swift-syntax/pull/1073

@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
https://github.com/apple/swift-syntax/pull/1073

@swift-ci smoke test

**Comment by drexin** [comment]
https://github.com/apple/swift-syntax/pull/1073

@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test macos

**Comment by drexin** [comment]
https://github.com/apple/swift-syntax/pull/1073

@swift-ci smoke test macos

**Comment by drexin** [comment]
https://github.com/apple/swift-syntax/pull/1073

@swift-ci smoke test macos

**Comment by drexin** [comment]
https://github.com/apple/swift-syntax/pull/1073

@swift-ci smoke test macos

**Comment by drexin** [comment]
https://github.com/apple/swift-syntax/pull/1073

@swift-ci smoke test macos

**Comment by drexin** [comment]
https://github.com/apple/swift-syntax/pull/1073

@swift-ci smoke test macos

File: lib/IRGen/ScalarTypeInfo.h [hunk 1]
```diff
@@ -289,7 +296,8 @@ class PODSingleScalarTypeInfo : public SingleScalarTypeInfo<Derived, Base> {
 
   TypeLayoutEntry *buildTypeLayoutEntry(IRGenModule &IGM,
                                         SILType T) const override {
-    return IGM.typeLayoutCache.getOrCreateScalarEntry(asDerived(), T);
+    return IGM.typeLayoutCache.getOrCreateScalarEntry(asDerived(), T,
```
Comment (aschwaighofer) [inline lib/IRGen/ScalarTypeInfo.h:299]: [MUST FIX] Missing check.

File: lib/IRGen/TypeLayout.cpp [hunk 2]
```diff
@@ -731,14 +783,168 @@ llvm::Value *ScalarTypeLayoutEntry::isBitwiseTakable(IRGenFunction &IGF) const {
       IGF.IGM.Int1Ty, typeInfo.isBitwiseTakable(ResilienceExpansion::Maximal));
 }
 
+llvm::Optional<std::vector<uint8_t>>
+ScalarTypeLayoutEntry::layoutString(IRGenModule &IGM) const {
+  switch (scalarKind) {
+  case ScalarKind::POD: {
+    assert(typeInfo.isFixedSize());
+    Size size = cast<FixedTypeInfo>(typeInfo).getFixedSize();
+    switch (size.getValueInBits()) {
```
Comment (aschwaighofer) [inline lib/IRGen/TypeLayout.cpp:792]: This does not handle i1 (bool).

File: stdlib/toolchain/CompatibilityBytecodeLayouts/BytecodeLayouts.h [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+
+#include "swift/Runtime/Metadata.h"
+#include <cstdint>
+#include <vector>
+
+// Layouts
+//
+// enum class LayoutType: char {
+//   // Scalars
+//   I8 = 'c',
```
Comment (aschwaighofer) [inline stdlib/toolchain/CompatibilityBytecodeLayouts/BytecodeLayouts.h:29]: Missing (at least) i1.

Comment (aschwaighofer) [reply stdlib/toolchain/CompatibilityBytecodeLayouts/BytecodeLayouts.h:29]: I don't think the scalar encoding as is is sufficient. LLVM builtin types are available -- eventually, we need to be able to match/encode the results of `getSpareBitsForType`.

**Review by aschwaighofer — CHANGES_REQUESTED** [review]
LGTM (mostly, besides on MUSTFIX comment)

File: lib/IRGen/ScalarTypeInfo.h (hunk 1 above)
Comment (drexin) [reply lib/IRGen/ScalarTypeInfo.h:302]: Fixed

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
https://github.com/apple/swift-syntax/pull/1073

@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test

**Pull request merged**